### ISSUE TITLE
Support for ShortText

### ISSUE THREAD

**Issue by Qqwy** [ExtremaIS/ttc-haskell#5]
For `ByteString` (8 words header, raw bytes, zero-copy-slicing, pinned, strict and lazy versions) 
there is `ShortByteString` (4 words header,raw bytes, no zero-copy slicing, not pinned, newtype of ByteArray).

For `Text` (6 words header, UTF8, zero-copy-slicing, strict and lazy versions)
there is `ShortText` (4 words header, UTF8, no zero-copy-slicing, newtype of ByteArray).

`ShortText` lives in the separate [`text-short` package](https://hackage.haskell.org/package/text-short), but I think it makes a lot of sense to support it as part of TTC.

<sub>EDIT: An earlier version of this message mistakenly referred to header sizes as e.g. '8 bytes' where it really is '8 words'.</sub>

**Comment by TravisCardwell** [comment]
Hi @Qqwy,

Thank you for the issue!

I have not included the `text-short` package because doing so would add a number of transitive dependencies.  I will investigate this again now.

On the current `develop` branch (`91da0c8d`), I checked the current transitive dependencies using `cabal freeze` as follows:

```
$ cabal clean
$ cabal configure --with-ghc ghc-9.6.3
$ cabal freeze --with-ghc ghc-9.6.3 --disable-tests --disable-benchmarks
```

Here are the resulting dependencies:

* `array`
* `base`
* `binary`
* `bytestring`
* `containers`
* `deepseq`
* `ghc-bignum`
* `ghc-boot-th`
* `ghc-prim`
* `pretty`
* `rts`
* `template-haskell`
* `text`

<details>
<summary>Dependency tree generated using <tt>cabal-plan</tt></summary>

```
ttc-1.3.0.0
 ├─ base-4.18.1.0
 │   ├─ ghc-bignum-1.3
 │   │   └─ ghc-prim-0.10.0
 │   │       └─ rts-1.0.2
 │   ├─ ghc-prim-0.10.0 ┄┄
 │   └─ rts-1.0.2 ┄┄
 ├─ bytestring-0.11.5.2
 │   ├─ base-4.18.1.0 ┄┄
 │   ├─ deepseq-1.4.8.1
 │   │   ├─ array-0.5.5.0
 │   │   │   └─ base-4.18.1.0 ┄┄
 │   │   ├─ base-4.18.1.0 ┄┄
 │   │   └─ ghc-prim-0.10.0 ┄┄
 │   ├─ ghc-prim-0.10.0 ┄┄
 │   └─ template-haskell-2.20.0.0
 │       ├─ base-4.18.1.0 ┄┄
 │       ├─ ghc-boot-th-9.6.3
 │       │   └─ base-4.18.1.0 ┄┄
 │       ├─ ghc-prim-0.10.0 ┄┄
 │       └─ pretty-1.1.3.6
 │           ├─ base-4.18.1.0 ┄┄
 │           ├─ deepseq-1.4.8.1 ┄┄
 │           └─ ghc-prim-0.10.0 ┄┄
 ├─ template-haskell-2.20.0.0 ┄┄
 └─ text-2.0.2
     ├─ array-0.5.5.0 ┄┄
     ├─ base-4.18.1.0 ┄┄
     ├─ binary-0.8.9.1
     │   ├─ array-0.5.5.0 ┄┄
     │   ├─ base-4.18.1.0 ┄┄
     │   ├─ bytestring-0.11.5.2 ┄┄
     │   └─ containers-0.6.7
     │       ├─ array-0.5.5.0 ┄┄
     │       ├─ base-4.18.1.0 ┄┄
     │       ├─ deepseq-1.4.8.1 ┄┄
     │       └─ template-haskell-2.20.0.0 ┄┄
     ├─ bytestring-0.11.5.2 ┄┄
     ├─ deepseq-1.4.8.1 ┄┄
     ├─ ghc-prim-0.10.0 ┄┄
     └─ template-haskell-2.20.0.0 ┄┄
```

</details>

One thing that I like about the current dependency set is that all of the transitive dependencies are GHC boot libraries.  The versions of these transitive dependencies are (generally) determined by the version of GHC being used.  When a new version of GHC is released, I can test and add support very quickly.

With `text-short` added as a dependency, I checked the transitive dependencies in the same way:

```
$ cabal clean
$ cabal configure --with-ghc ghc-9.6.3
$ cabal freeze --with-ghc ghc-9.6.3 --disable-tests --disable-benchmarks
```

Here are the resulting dependencies:

* `array`
* `base`
* `binary`
* `bytestring`
* `containers`
* `deepseq`
* `exceptions`
* `filepath`
* `ghc-bignum`
* `ghc-boot-th`
* `ghc-prim`
* `hashable`
* `mtl`
* `pretty`
* `rts`
* `stm`
* `template-haskell`
* `text`
* `text-short`
* `transformers`

<details>
<summary>Dependency tree generated using <tt>cabal-plan</tt></summary>

```
ttc-1.3.0.0
 ├─ base-4.18.1.0
 │   ├─ ghc-bignum-1.3
 │   │   └─ ghc-prim-0.10.0
 │   │       └─ rts-1.0.2
 │   ├─ ghc-prim-0.10.0 ┄┄
 │   └─ rts-1.0.2 ┄┄
 ├─ bytestring-0.11.5.2
 │   ├─ base-4.18.1.0 ┄┄
 │   ├─ deepseq-1.4.8.1
 │   │   ├─ array-0.5.5.0
 │   │   │   └─ base-4.18.1.0 ┄┄
 │   │   ├─ base-4.18.1.0 ┄┄
 │   │   └─ ghc-prim-0.10.0 ┄┄
 │   ├─ ghc-prim-0.10.0 ┄┄
 │   └─ template-haskell-2.20.0.0
 │       ├─ base-4.18.1.0 ┄┄
 │       ├─ ghc-boot-th-9.6.3
 │       │   └─ base-4.18.1.0 ┄┄
 │       ├─ ghc-prim-0.10.0 ┄┄
 │       └─ pretty-1.1.3.6
 │           ├─ base-4.18.1.0 ┄┄
 │           ├─ deepseq-1.4.8.1 ┄┄
 │           └─ ghc-prim-0.10.0 ┄┄
 ├─ template-haskell-2.20.0.0 ┄┄
 ├─ text-2.0.2
 │   ├─ array-0.5.5.0 ┄┄
 │   ├─ base-4.18.1.0 ┄┄
 │   ├─ binary-0.8.9.1
 │   │   ├─ array-0.5.5.0 ┄┄
 │   │   ├─ base-4.18.1.0 ┄┄
 │   │   ├─ bytestring-0.11.5.2 ┄┄
 │   │   └─ containers-0.6.7
 │   │       ├─ array-0.5.5.0 ┄┄
 │   │       ├─ base-4.18.1.0 ┄┄
 │   │       ├─ deepseq-1.4.8.1 ┄┄
 │   │       └─ template-haskell-2.20.0.0 ┄┄
 │   ├─ bytestring-0.11.5.2 ┄┄
 │   ├─ deepseq-1.4.8.1 ┄┄
 │   ├─ ghc-prim-0.10.0 ┄┄
 │   └─ template-haskell-2.20.0.0 ┄┄
 └─ text-short-0.1.5
     ├─ base-4.18.1.0 ┄┄
     ├─ binary-0.8.9.1 ┄┄
     ├─ bytestring-0.11.5.2 ┄┄
     ├─ deepseq-1.4.8.1 ┄┄
     ├─ ghc-prim-0.10.0 ┄┄
     ├─ hashable-1.4.3.0
     │   ├─ base-4.18.1.0 ┄┄
     │   ├─ bytestring-0.11.5.2 ┄┄
     │   ├─ containers-0.6.7 ┄┄
     │   ├─ deepseq-1.4.8.1 ┄┄
     │   ├─ filepath-1.4.100.4
     │   │   ├─ base-4.18.1.0 ┄┄
     │   │   ├─ bytestring-0.11.5.2 ┄┄
     │   │   ├─ deepseq-1.4.8.1 ┄┄
     │   │   ├─ exceptions-0.10.7
     │   │   │   ├─ base-4.18.1.0 ┄┄
     │   │   │   ├─ mtl-2.3.1
     │   │   │   │   ├─ base-4.18.1.0 ┄┄
     │   │   │   │   └─ transformers-0.6.1.0
     │   │   │   │       └─ base-4.18.1.0 ┄┄
     │   │   │   ├─ stm-2.5.1.0
     │   │   │   │   ├─ array-0.5.5.0 ┄┄
     │   │   │   │   └─ base-4.18.1.0 ┄┄
     │   │   │   ├─ template-haskell-2.20.0.0 ┄┄
     │   │   │   └─ transformers-0.6.1.0 ┄┄
     │   │   └─ template-haskell-2.20.0.0 ┄┄
     │   ├─ ghc-bignum-1.3 ┄┄
     │   ├─ ghc-prim-0.10.0 ┄┄
     │   └─ text-2.0.2 ┄┄
     ├─ template-haskell-2.20.0.0 ┄┄
     └─ text-2.0.2 ┄┄
```

</details>

The additional dependencies are as follows:

* `exceptions` (boot)
* `filepath` (boot)
* `hashable`
* `mtl` (boot)
* `stm` (boot)
* `text-short`
* `transformers` (boot)

I feel that this is a significant addition to the dependencies.  It looks like the decision to provide a `Hashable` instance for `ShortText` is what pulls the `filepath` dependency, which brings along `exceptions`, `mtl`, `transformers`, and `stm`.  In particular, I have had frustrations with `mtl` transitive dependencies and see it as a significant addition.  That said, most projects include all of these transitive dependencies anyway, so I would consider adding them if there is a need to do so.

Personally, I have used `ShortByteString` in a server where pinned `ByteString` values had caused memory fragmentation that prevented the server from running indefinitely.  Since `Text` is unpinned, my understanding is that the raison d'être of `ShortText` is the small size, which is much less of a benefit now that `Text` also uses the UTF-8 encoding.  I noticed that `aeson` has a `text-short` dependency from version 2, though, so perhaps usage of it is increasing...

What do you think?  Do you have a need for supporting `ShortText` in a current project?

**Comment by Qqwy** [comment]
The desire to only depend on boot libraries is very understandable. However, the design of `Textual` requires that support is added as part of `ttc`; it cannot be added in `text-short` or as (shudder) an orphan.

One possibility could be to add a configuration flag to `ttc` to conditionally support ShortText without pulling in the dependencies by default.

What do you think?

**Comment by TravisCardwell** [comment]
Hi @Qqwy,

I pushed a commit to the [`text-short` branch](https://github.com/ExtremaIS/ttc-haskell/tree/text-short), in case you are working on a project that uses `ShortText` and would like to test.

Using a configuration flag to enable support for `text-short` is an interesting idea.  I think it would be better to (unconditionally) add support, however, as gating API differences with a configuration flag can cause complications in packages that depend on the configured package.

I am happy to add support, but I prefer to make such a decision with data to back it up.  I would like to confirm that people actually use `text-short` (not just provide instances or conversion functions) now that `Text` uses UTF-8.  I did a quick survey of Hackage packages, and almost all concrete uses are by the same author.  Searching for `"Data.Text.Short" language:Haskell` on GitHub  yields *many* results, however!

Also, I noticed that the difference in memory footprint between `Text` and `ShortText` is much more significant than indicated above.  The `ShortText` type saves two *words*, not *bytes*!

If the above commit looks good, I will schedule to merge to `develop` and release in `ttc-1.4.0.0`.

Thank you again!

**Comment by Qqwy** [comment]
> Using a configuration flag to enable support for text-short is an interesting idea. I think it would be better to (unconditionally) add support, however, as gating API differences with a configuration flag can cause complications in packages that depend on the configured package.

Yeah, this is an idiom which has become quite common in Rust to reduce the amount of required dependencies, though I believe that Rust's package manager (`cargo`) does some clever 'feature flag resolution' tricks to make this ergonomic in a way which `cabal`/`stack` currently do not in Haskell.

> I am happy to add support, but I prefer to make such a decision with data to back it up.

👍

>  I did a quick survey of Hackage packages, and almost all concrete uses are by the same author. Searching for "Data.Text.Short" language:Haskell on GitHub yields many results, however!

Wow, that is very interesting!

> 

Also, I noticed that the difference in memory footprint between Text and ShortText is much more significant than indicated above. The ShortText type saves two words, not bytes!

Whoops! I wrote bytes where I meant words. I've edited the first message to make this clearer.

---

Thank you for publishing a branch to try out! The project I am currently working on (which is how I encountered this issue) is the [symbolize](https://hackage.haskell.org/package/symbolize) library. 

At first I was using TTC's `Textual` class using `ShortByteString` internally, but because I wanted to make sure that correct unicode is always enforced (which `TTC`'s typeclass intentionally only does when you actually convert to a `Text`-like type, I wanted to switch to `ShortText`. There are two reasons to use `ShortText` over `Text` in this particular scenario:
- The afore-mentioned 2 words memory saving per entry;
- `Text`'s slicing is actually dangerous here as it would result in memory leaks: If you would store e.g. the word `hello` when it were to originate from a 100MB text document, you'd end up with a reference to this full 100MB in memory.

For the time being I have extracted the functionality which I needed from `Data.TTC.Textual` into a bespoke typeclass (also called `Textual`, because it is a really good name).

For a symbol table I actually think conversion _has_ to go through a single type (or rather: the symbol table itself should store a single compact representation, and therefore all reads/writes neccesarily need to convert to/from that representation) so I don't end up relying on that particular feature of `Data.TTC.Textual`.

... which finally has me reconsider whether `Symbolize` should depend on `TTC` or not, even when `ShortText`-support is merged. But regardless, I still think that support for `ShortText` would be a very good feature to add to this library. 
The `text-short` branch looks very promising!

**Comment by TravisCardwell** [comment]
Hi @Qqwy,

Your `symbolize` project is interesting, and `ShortText` is indeed a good choice of type to use!

TTC assumes that text is valid Unicode.  The `Textual` methods are all total because they use lenient decoding that replaces invalid bytes with Unicode replacement characters (`U+FFFD`).  One can check for invalid Unicode in a `Parse` instance, however.  For example, your symbol table uses the `ShortText` type, but it could instead use a `newtype` wrapper and instances like the following:

```haskell
newtype SymbolText = SymbolText { unSymbolText :: ShortText }
  deriving Show

instance TTC.Parse SymbolText where
  parse = TTC.asST $ \st -> do
    when (isJust $ ST.find (== '\xfffd') st) . Left $ TTC.fromS "invalid Unicode"
    pure $ SymbolText st

instance TTC.Render SymbolText where
  render = TTC.fromST . unSymbolText
```

This (untested) example can be tested as follows:

```haskell
print (TTC.parseBS "test\xe3" :: Either String SymbolText)
```

You could implement further restrictions on symbol syntax in the `Parse` instance if desired.  For example, you could use a fold to both limit the character set and check for invalid Unicode, providing good error messages, in a single pass.

The symbol text is always stored as `ShortText`; the `Textual` parameters and return values just facilitate conversion to/from all of the `Textual` types.

If you would like to instead use [`decodeUtf8`][] (in order to get the `UnicodeException` or for better performance if there are no other restrictions), then TTC may not be a good fit, as it would need to be done outside of TTC instances.  In this case, you might want to check out Taylor Fausak's [`witch`][] library.  It does not have built-in support for `text-short` either, though.

[`decodeUtf8`]: <https://hackage.haskell.org/package/text-2.1/docs/Data-Text-Encoding.html#v:decodeUtf8-39->
[`witch`]: <https://hackage.haskell.org/package/witch>

No worries if you implement what you need inside of your library instead of using TTC.  You have convinced me that `text-short` support should be added regardless.  Thanks!

**Comment by TravisCardwell** [comment]
@Qqwy I released [`ttc-1.4.0.0`][] with `ShortText` support.  I will keep the [`text-short` branch][] for a while in case you are using it, but I will eventually remove it after some time.

[`ttc-1.4.0.0`]: <https://hackage.haskell.org/package/ttc-1.4.0.0>
[`text-short` branch]: <https://github.com/ExtremaIS/ttc-haskell/tree/text-short>

**Comment by Qqwy** [comment]
Thank you very much!
